### PR TITLE
Migrate pytorch_python_doc_build to github action

### PR DESCRIPTION
**Pull request by zhouzhuojie**
# Changes

This PR migrates `pytorch_python_doc_build` from circleci to github actions.

Noticeable changes
- Refactor `docker cp` into a single `docker run` with volume mount, because the in circleci volume is not accessible from its remote docker engine
- `pytorch_python_doc_push` job will have a race condition with circleci, which will be migrated in separate PRs

### CODE REVIEW

File: .github/workflows/doc_build.yml
Comment (seemethere) [inline .github/workflows/doc_build.yml:50]: I think it might be better to make this a part of the regular linux CI workflow template and just have it gated like so:
```
if: {{ env.BUILD_ENVIRONMENT == "pytorch-linux-xenial-py3.6-gcc5.4"
```

It'd make it easier to grab the artifacts etc.

File: .github/workflows/doc_build.yml
Comment (zhouzhuojie) [inline .github/workflows/doc_build.yml:9]: this line `pull_request` will just for testing purpose - really need to trigger this workflow to start iterating on the testing, ideally the doc build should only run on master and release after `Linux CI (pytorch-linux-xenial-py3.6-gcc5.4)` is finished.

File: .github/workflows/doc_build.yml
Comment (zhouzhuojie) [reply .github/workflows/doc_build.yml:50]: makes sense, the artifact upload/download looks like can only be shared within the same workflow.

File: .github/workflows/pytorch-linux-xenial-py3.6-gcc5.4.yml
Comment (zhouzhuojie) [inline .github/workflows/pytorch-linux-xenial-py3.6-gcc5.4.yml:243]: here's a job `if` condition that `doc_push` only runs on master

**Comment by zhouzhuojie** [comment]
@seemethere updated, and moved to the same workflow as the previous python CI workflow so that they can share the artifacts. May need your approval again for the new workflows

**Comment by zhouzhuojie** [comment]
Oh, there's a template to generate this, hold on, will move the code there

**Comment by zhouzhuojie** [comment]
> Oh, there's a template to generate this, hold on, will move the code there

Migrated to linux_ci_workflow.yml.in, and decouples doc build from tightly coupled with `pytorch-linux-xenial-py3.6-gcc5.4`, and instead just make it as a template rendering option `enable_doc_jobs`

File: .github/templates/linux_ci_workflow.yml.in
Comment (zhouzhuojie) [inline .github/templates/linux_ci_workflow.yml.in:181]: we can also move cpp docs job to here in the future

File: .github/templates/linux_ci_workflow.yml.in
Comment (seemethere) [inline .github/templates/linux_ci_workflow.yml.in:183]: This needs to run on a `linux.2xlarge` in order to have access to ECR

Comment (zhouzhuojie) [reply .github/templates/linux_ci_workflow.yml.in]: I see, updated

File: .github/workflows/pytorch-linux-xenial-py3.6-gcc5.4.yml
Comment (zhouzhuojie) [inline .github/workflows/pytorch-linux-xenial-py3.6-gcc5.4.yml:244]: Note that the secrets GITHUB_PYTORCHBOT_TOKEN needs to be set up first before merging

**Comment by samestep** [comment]
question: why is the docs build being added as a gated job inside the `.github/templates/linux_ci_workflow.yml.in` file, rather than as its own new template?

**Comment by samestep** [comment]
also, I have enabled GitHub Actions to run on this PR; when you have the chance, could you fix the lints? https://github.com/pytorch/pytorch/runs/2504623867

**Comment by samestep** [comment]
> currently hidden by the ShellCheck failure, but i'm guessing it would be failing

@pytorch/pytorch-dev-infra I am currently working on #57572 which should resolve this ergonomics issue

**Comment by zhouzhuojie** [comment]
> question: why is the docs build being added as a gated job inside the `.github/templates/linux_ci_workflow.yml.in` file, rather than as its own new template? I guess it's just because it depends on having the build already done?
> 
> (cc @seemethere)

Originally, I was implementing it in its own workflow, later I was debugging it and I agreed with @seemethere's comment that we can put the docs build together with linux_ci_workflow.yml.in, maybe they can share more steps together.

But, the only dependencies that the doc build job requires is a `calculate-docker-image` to figure out the target docker image, other build artifacts like the built docs repo can be encapsulated into the dedicated docs related workflow.

I can refactor this and make it cleaner.

File: .github/workflows/docs.yml
Comment (zhouzhuojie) [inline .github/workflows/docs.yml:94]: TODO, this may also need to include the cases for
```
- refs/heads/master
- refs/tags/vx.y.z
- refs/tags/vx.y.z-rc
```

**Comment by samestep** [comment]
> Originally, I was implementing it in its own workflow, later I was debugging it and I agreed with @seemethere's comment that we can put the docs build together with linux_ci_workflow.yml.in, maybe they can share more steps together.

makes sense; thanks!

> But, the only dependencies that the doc build job requires is a `calculate-docker-image` to figure out the target docker image, other build artifacts like the built docs repo can be encapsulated into the dedicated docs related workflow.

either way sounds good to me :)

File: .github/templates/linux_ci_workflow.yml.in
Comment (zhouzhuojie) [reply .github/templates/linux_ci_workflow.yml.in]: yes, did that, the only reason `docker cp` is necessary in circleci is that their docker engine is wired remotely - https://circleci.com/docs/2.0/building-docker-images/#mounting-folders. While github action docker is running locally and volume mount is doable.

File: .github/templates/linux_ci_workflow.yml.in
Comment (zhouzhuojie) [reply .github/templates/linux_ci_workflow.yml.in]: thanks, yes sure, I may need to figure out how to iterate this faster

every change requires approvals - so each iteration requires a few hours of wait - anything special I can do to test the flow e2e on my own before requiring your review time?

Comment (samestep) [reply .github/templates/linux_ci_workflow.yml.in]: yeah this is unfortunately a pain point that we haven't yet solved; one hacky solution we could take in the short-term is have you make a tiny cosmetic diff and Jedi-land it so that GitHub no longer thinks you're a "first-time contributor", after which (I hope) you wouldn't need approval to run CI on this PR; @seemethere what do you think?

Comment (samestep) [reply .github/templates/linux_ci_workflow.yml.in]: as for running locally, I don't think we have a good solution for that yet either (but I could be wrong)

File: .github/workflows/cancel_redundant_workflows.yml
Comment (zhouzhuojie) [inline .github/workflows/cancel_redundant_workflows.yml:10]: generated from .github/scripts/regenerate_cancel_redundant_workflow.py

File: .github/workflows/docs.yml
Comment (walterddr) [reply .github/workflows/docs.yml:94]: please add a todo comment 
```
# TODO: add tags doc push for regex: `v[0-9]+(\.[0-9]+)*-rc[0-9]+`
```

also currently doc push is only enabled on nightly instead of master, should we change this to nightly?

File: .github/templates/linux_ci_workflow.yml.in
Comment (zhouzhuojie) [reply .github/templates/linux_ci_workflow.yml.in]: tested a few times, and yes, indeed it requires the artifacts generated from the `build` step, I updated and may need a new approval for the workflows.

Noticeable changes are
- refactor `docker cp` into a single `docker run`
- download artifacts from `build` step and mount as a volume
- change from `sh` to `bash` to run in docker as it requires `pushd` which is a bash-only command
- refactor circleci tag into GITHUB_REF - as gha only supports git ref

Comment (zhouzhuojie) [reply .github/templates/linux_ci_workflow.yml.in]: To pass the "first-time contributor" PR https://github.com/pytorch/pytorch/pull/57779

**Comment by samestep** [comment]
now that #57779 is about to be merged, _hopefully_ it should automatically run GHA workflows for you without someone else having to approve them... if not then I'll be sad haha

File: .github/templates/linux_ci_workflow.yml.in
Comment (seemethere) [inline .github/templates/linux_ci_workflow.yml.in:281]: I think we probably don't need to do this here, the reason we were doing the whole `docker commit && docker push` was to preserve artifacts from one stage to another

Comment (zhouzhuojie) [reply .github/templates/linux_ci_workflow.yml.in:281]: This is for feature parity from circleci jobs. I think it can be useful if people rely on this debug image for quick iteration on their docs. Should we still remove it?

**Review by samestep — APPROVED** [review]

**Pull request closed without merging**
